
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.3.2">
    
    
      
        <title>random_sampling - Sequence composition virus-host matching project</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.24c991ad.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.3f5d1f46.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    
      


    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="teal">
  
    
    <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
      <script>var palette=__get("__palette");if(null!==palette&&"object"==typeof palette.color)for(var key in palette.color)document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#random_sampling_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Sequence composition virus-host matching project" class="md-header__button md-logo" aria-label="Sequence composition virus-host matching project" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Sequence composition virus-host matching project
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              random_sampling
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="teal"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="orange" data-md-color-accent="orange"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/phenolophthaleinum/fastDNA-faiss/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    phenolophthaleinum/fastDNA-faiss
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href="../.." class="md-tabs__link">
      Main page
    </a>
  </li>

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../make-model/" class="md-tabs__link md-tabs__link--active">
        Reference
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Sequence composition virus-host matching project" class="md-nav__button md-logo" aria-label="Sequence composition virus-host matching project" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Sequence composition virus-host matching project
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/phenolophthaleinum/fastDNA-faiss/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    phenolophthaleinum/fastDNA-faiss
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Main page
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../make-model/" class="md-nav__link">
        make-model
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          random_sampling
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        random_sampling
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#random_sampling" class="md-nav__link">
    random_sampling
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random_sampling.main_procedure" class="md-nav__link">
    main_procedure()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random_sampling.sample_sequences" class="md-nav__link">
    sample_sequences()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../workflow2/" class="md-nav__link">
        workflow2
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#random_sampling" class="md-nav__link">
    random_sampling
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random_sampling.main_procedure" class="md-nav__link">
    main_procedure()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random_sampling.sample_sequences" class="md-nav__link">
    sample_sequences()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/phenolophthaleinum/fastDNA-faiss/blob/master/docs/reference/random_sampling.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="random_sampling_1">random_sampling</h1>


  <div class="doc doc-object doc-module">

<a id="random_sampling"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">








  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="random_sampling.main_procedure">
<code class="highlight language-python"><span class="n">main_procedure</span><span class="p">(</span><span class="n">wd</span><span class="p">,</span> <span class="n">host</span><span class="p">,</span> <span class="n">virus</span><span class="p">,</span> <span class="n">full</span><span class="p">,</span> <span class="n">host_dir</span><span class="p">,</span> <span class="n">virus_dir</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n_vir_samples</span><span class="p">,</span> <span class="n">n_host_samples</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Main function of the module that samples genomes randomly:</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>wd</code></td>
        <td><code>str</code></td>
        <td><p>Working directory path</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>host</code></td>
        <td><code>bool</code></td>
        <td><p>Flag, which allows sampling only host genomes</p>
<div class="admonition p-obsolete">
<p class="admonition-title">Obsolete</p>
<p>In current vision of the whole process, this parameter is obsolete. Using it might be undesirable and cause program to fail. Might be removed in later version.</p>
</div></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>virus</code></td>
        <td><code>bool</code></td>
        <td><p>Flag, which allows sampling only virus genomes.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>full</code></td>
        <td><code>bool</code></td>
        <td><p>Flag, which allows sampling both virus and host genomes.</p>
<div class="admonition p-obsolete">
<p class="admonition-title">Obsolete</p>
<p>In current vision of the whole process, this parameter is obsolete. Using it might be undesirable and cause program to fail. Might be removed in later version.</p>
</div></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>host_dir</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory where sampled host fragments will be saved</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>virus_dir</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory where sampled virus fragments will be saved</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>length</code></td>
        <td><code>int</code></td>
        <td><p>Length of sampled fragments</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>n_vir_samples</code></td>
        <td><code>int</code></td>
        <td><p>Number of samples to take from a virus genome</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>n_host_samples</code></td>
        <td><code>int</code></td>
        <td><p>Number of samples to take from a host genome</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>n_nuc_threshold</code></td>
        <td><code>float</code></td>
        <td><p>Maximum allowed ambiguous nucleotide content threshold in sampled sequences (in percents)</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\random_sampling.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">main_procedure</span><span class="p">(</span><span class="n">wd</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">host</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">virus</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">full</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">host_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">virus_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                   <span class="n">n_vir_samples</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">n_host_samples</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Main function of the module that samples genomes randomly:</span>

<span class="sd">    Args:</span>
<span class="sd">        wd: Working directory path</span>
<span class="sd">        host: Flag, which allows sampling only host genomes</span>
<span class="sd">            !!! p-obsolete &quot;Obsolete&quot;</span>
<span class="sd">                In current vision of the whole process, this parameter is obsolete. Using it might be undesirable and cause program to fail. Might be removed in later version.</span>
<span class="sd">        virus: Flag, which allows sampling only virus genomes.</span>
<span class="sd">        full: Flag, which allows sampling both virus and host genomes.</span>
<span class="sd">            !!! p-obsolete &quot;Obsolete&quot;</span>
<span class="sd">                In current vision of the whole process, this parameter is obsolete. Using it might be undesirable and cause program to fail. Might be removed in later version.</span>
<span class="sd">        host_dir: Path to directory where sampled host fragments will be saved</span>
<span class="sd">        virus_dir: Path to directory where sampled virus fragments will be saved</span>
<span class="sd">        length: Length of sampled fragments</span>
<span class="sd">        n_vir_samples: Number of samples to take from a virus genome</span>
<span class="sd">        n_host_samples: Number of samples to take from a host genome</span>
<span class="sd">        n_nuc_threshold: Maximum allowed ambiguous nucleotide content threshold in sampled sequences (in percents)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># colorama</span>
    <span class="n">init</span><span class="p">()</span>

    <span class="c1"># timer</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">timer</span><span class="p">()</span>

    <span class="c1"># importing json data about hosts</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">get_host_data</span><span class="p">()</span>

    <span class="n">final_records</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># parallel sampling records of all files and dumping them into one file</span>
    <span class="k">if</span> <span class="n">host</span><span class="p">:</span>
        <span class="n">new_records</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">sample_sequences</span><span class="p">)(</span><span class="n">file</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n_host_samples</span><span class="p">,</span> <span class="n">wd</span><span class="p">,</span> <span class="n">virus</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span>
            <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">host_dir</span><span class="si">}</span><span class="s2">*.fna&quot;</span><span class="p">))</span>

        <span class="c1"># for host</span>
        <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">new_records</span><span class="p">:</span>
            <span class="n">final_records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sublist</span><span class="p">)</span>

        <span class="c1"># host_file = Path(f&quot;{wd}host/samples/host_samples.fasta&quot;)</span>
        <span class="c1"># if host_file.exists():</span>
        <span class="c1">#     os.system(f&quot;{wd}host/samples/host_samples.fasta&quot;)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/samples/host_samples.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w_fh</span><span class="p">:</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">final_records</span><span class="p">,</span> <span class="n">w_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="c1"># mapping samples to nbci ids and dumping them into a file; edit 10.11.21 - much faster</span>
        <span class="n">p_records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/samples/host_samples.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
        <span class="n">ids_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">p_records</span><span class="p">]</span>
        <span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ids_records</span><span class="p">):</span>
            <span class="n">d</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
        <span class="c1"># for id in set(ids_records):</span>
        <span class="c1">#     keys = [index for index, value in enumerate(ids_records) if value == id]</span>
        <span class="c1">#     for key in keys:</span>
        <span class="c1">#         d[key] = id</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/maps/sample_map.json&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">fh</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">virus</span><span class="p">:</span>
        <span class="n">new_records</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">sample_sequences</span><span class="p">)(</span><span class="n">file</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n_vir_samples</span><span class="p">,</span> <span class="n">wd</span><span class="p">,</span> <span class="n">virus</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span>
            <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">virus_dir</span><span class="si">}</span><span class="s2">*.fna&quot;</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">full</span><span class="p">:</span>
        <span class="n">new_records</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">sample_sequences</span><span class="p">)(</span><span class="n">file</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n_host_samples</span><span class="p">,</span> <span class="n">wd</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">,</span> <span class="n">virus</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span>
            <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">host_dir</span><span class="si">}</span><span class="s2">*.fna&quot;</span><span class="p">))</span>

        <span class="c1"># for host</span>
        <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">new_records</span><span class="p">:</span>
            <span class="n">final_records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sublist</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/samples/host_samples.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w_fh</span><span class="p">:</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">final_records</span><span class="p">,</span> <span class="n">w_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="c1"># mapping samples to nbci ids and dumping them into a file; edit 10.11.21 - much faster</span>
        <span class="n">p_records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/samples/host_samples.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
        <span class="n">ids_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">p_records</span><span class="p">]</span>
        <span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ids_records</span><span class="p">):</span>
            <span class="n">d</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
        <span class="c1"># for id in set(ids_records):</span>
        <span class="c1">#     keys = [index for index, value in enumerate(ids_records) if value == id]</span>
        <span class="c1">#     for key in keys:</span>
        <span class="c1">#         d[key] = id</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">host/maps/sample_map.json&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">fh</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

        <span class="n">new_records</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">sample_sequences</span><span class="p">)(</span><span class="n">file</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n_vir_samples</span><span class="p">,</span> <span class="n">wd</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">,</span> <span class="n">virus</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span>
            <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">virus_dir</span><span class="si">}</span><span class="s2">*.fna&quot;</span><span class="p">))</span>

    <span class="c1"># for virus, but second slower</span>
    <span class="c1"># for sublist in new_records:</span>
    <span class="c1">#     with open(f&quot;X:/edwards2016/virus/samples/{sublist[0].id}_samples.fasta&quot;, &quot;a&quot;) as w_fh:</span>
    <span class="c1">#         SeqIO.write(sublist, w_fh, &quot;fasta&quot;)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">timer</span><span class="p">()</span>
    <span class="n">runtime</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="n">start</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Fore</span><span class="o">.</span><span class="n">GREEN</span><span class="si">}</span><span class="s2">[random-sampling] Done in </span><span class="si">{</span><span class="n">runtime</span><span class="si">:</span><span class="s2">.6f</span><span class="si">}</span><span class="s2"> seconds&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="random_sampling.sample_sequences">
<code class="highlight language-python"><span class="n">sample_sequences</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">wd</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">,</span> <span class="n">virus</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Function reads each given fasta file, randomly samples n subsequences of a defined length, creates new records with those subsequences and puts these new records into a list (which is finally returned).</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>file</code></td>
        <td><code>str</code></td>
        <td><p>Full path to a fasta file</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>wd</code></td>
        <td><code>str</code></td>
        <td><p>Working directory path</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>length</code></td>
        <td><code>int</code></td>
        <td><p>Length of a sampled subsequence</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>n</code></td>
        <td><code>int</code></td>
        <td><p>Number of subsequences to be sampled</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>n_nuc_threshold</code></td>
        <td><code>float</code></td>
        <td><p>Ambiguous nucleotide content threshold in sampled sequences (in percents)</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>virus</code></td>
        <td><code>bool</code></td>
        <td><p>Flag which enables or disables virus genomes sampling</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>list[SeqRecord]</code></td>
      <td><p>Returns list of newly created biopython <code>SeqRecord</code>, each contains sampled subsequences.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\random_sampling.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">sample_sequences</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">n</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">wd</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">n_nuc_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">virus</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Function reads each given fasta file, randomly samples n subsequences of a defined length, creates new records with those subsequences and puts these new records into a list (which is finally returned).</span>

<span class="sd">    Args:</span>
<span class="sd">        file (str): Full path to a fasta file</span>
<span class="sd">        wd (str): Working directory path</span>
<span class="sd">        length (int): Length of a sampled subsequence</span>
<span class="sd">        n (int): Number of subsequences to be sampled</span>
<span class="sd">        n_nuc_threshold (float): Ambiguous nucleotide content threshold in sampled sequences (in percents)</span>
<span class="sd">        virus (bool): Flag which enables or disables virus genomes sampling</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[SeqRecord]: Returns list of newly created biopython `SeqRecord`, each contains sampled subsequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_records</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="n">new_seq</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">pick</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">while</span> <span class="ow">not</span> <span class="n">valid</span><span class="p">:</span>
            <span class="n">pick</span> <span class="o">=</span> <span class="n">secrets</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="n">length</span><span class="p">))</span>
            <span class="n">new_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">pick</span><span class="p">:</span><span class="n">pick</span> <span class="o">+</span> <span class="n">length</span><span class="p">]))</span>
            <span class="n">n_content</span> <span class="o">=</span> <span class="n">new_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">n_content</span> <span class="o">/</span> <span class="n">length</span> <span class="o">&lt;</span> <span class="n">n_nuc_threshold</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Too high N content (</span><span class="si">{</span><span class="n">n_content</span> <span class="o">/</span> <span class="n">length</span><span class="si">}</span><span class="s2">%). Sampling again&quot;</span><span class="p">)</span>

        <span class="n">desc</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="si">}</span><span class="s1"> sample_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1"> gen_pos:(</span><span class="si">{</span><span class="n">pick</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">pick</span> <span class="o">+</span> <span class="n">length</span><span class="si">}</span><span class="s1">)&#39;</span>
        <span class="n">new_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
            <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="n">seq</span><span class="o">=</span><span class="n">new_seq</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="n">desc</span>
        <span class="p">)</span>
        <span class="n">new_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_record</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">virus</span><span class="p">:</span>
        <span class="c1"># if virus, this is possible</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s2">virus/samples/</span><span class="si">{</span><span class="n">new_records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_samples.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w_fh</span><span class="p">:</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_records</span><span class="p">,</span> <span class="n">w_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_records</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../make-model/" class="md-footer__link md-footer__link--prev" aria-label="Previous: make-model" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              make-model
            </div>
          </div>
        </a>
      
      
        
        <a href="../utils/" class="md-footer__link md-footer__link--next" aria-label="Next: utils" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              utils
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        
          Made with
          <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
            Material for MkDocs
          </a>
        
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.e99e714c.min.js", "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.bc35569b.min.js"></script>
      
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
      
    
  </body>
</html>