
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.2, mkdocs-material-7.3.2">
    
    
      
        <title>make-model - Sequence composition virus-host matching project</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.24c991ad.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.3f5d1f46.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    
      


    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="teal">
  
    
    <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
      <script>var palette=__get("__palette");if(null!==palette&&"object"==typeof palette.color)for(var key in palette.color)document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#make-model_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Sequence composition virus-host matching project" class="md-header__button md-logo" aria-label="Sequence composition virus-host matching project" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Sequence composition virus-host matching project
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              make-model
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="teal"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="orange" data-md-color-accent="orange"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/phenolophthaleinum/fastDNA-faiss/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    phenolophthaleinum/fastDNA-faiss
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href="../.." class="md-tabs__link">
      Main page
    </a>
  </li>

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="./" class="md-tabs__link md-tabs__link--active">
        Reference
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Sequence composition virus-host matching project" class="md-nav__button md-logo" aria-label="Sequence composition virus-host matching project" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Sequence composition virus-host matching project
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/phenolophthaleinum/fastDNA-faiss/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    phenolophthaleinum/fastDNA-faiss
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Main page
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          make-model
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        make-model
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#make-model" class="md-nav__link">
    make-model
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.debug_filtering" class="md-nav__link">
    debug_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.fasta_parallel" class="md-nav__link">
    fasta_parallel()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.hybrid_filtering" class="md-nav__link">
    hybrid_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.main_procedure" class="md-nav__link">
    main_procedure()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.no_filtering" class="md-nav__link">
    no_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.tax_filtering" class="md-nav__link">
    tax_filtering()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../random_sampling/" class="md-nav__link">
        random_sampling
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../utils/" class="md-nav__link">
        utils
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../workflow2/" class="md-nav__link">
        workflow2
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#make-model" class="md-nav__link">
    make-model
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.debug_filtering" class="md-nav__link">
    debug_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.fasta_parallel" class="md-nav__link">
    fasta_parallel()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.hybrid_filtering" class="md-nav__link">
    hybrid_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.main_procedure" class="md-nav__link">
    main_procedure()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.no_filtering" class="md-nav__link">
    no_filtering()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#make-model.tax_filtering" class="md-nav__link">
    tax_filtering()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/phenolophthaleinum/fastDNA-faiss/blob/master/docs/reference/make-model.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="make-model_1">make-model</h1>


  <div class="doc doc-object doc-module">

<a id="make-model"></a>
    <div class="doc doc-contents first">




  <div class="doc doc-children">













  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.debug_filtering">
<code class="highlight language-python"><span class="n">debug_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Edwards' dataset of host is filtered to one, specific taxonomy level and from it, all host genomes that represent
this taxonomy level are chosen.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>filter</code></td>
        <td><code>str</code></td>
        <td><p>A flag to filter through this function (TODO: this is not final, this will act as an actual filter value in the future)</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>Tuple[List[str], List[str]]</code></td>
      <td><p>A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">debug_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;Edwards&#39; dataset of host is filtered to one, specific taxonomy level and from it, all host genomes that represent</span>
<span class="sd">    this taxonomy level are chosen.</span>

<span class="sd">    Args:</span>
<span class="sd">        filter: A flag to filter through this function (TODO: this is not final, this will act as an actual filter value in the future)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[List[str], List[str]]: A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># filtering all hosts by a chosen level</span>
    <span class="n">filter_host</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">target</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Clostridiales&quot;</span><span class="p">]</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">target</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">host_data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="n">all_tax_levels</span><span class="p">[</span><span class="nb">filter</span><span class="p">][</span><span class="s1">&#39;order&#39;</span><span class="p">]]</span> <span class="o">==</span> <span class="n">name</span><span class="p">:</span>
                <span class="n">filenames</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">host</span><span class="p">)</span>

    <span class="c1"># list of paths to be returned</span>
    <span class="n">filepaths</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;HOST&#39;</span><span class="p">][</span><span class="s1">&#39;host_genomes&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">host</span><span class="si">}</span><span class="s2">.fna&quot;</span> <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>

    <span class="c1"># flatten if needed</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filenames</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">filenames</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">temp_files_path</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">filepaths</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="n">temp_files</span>
        <span class="n">filepaths</span> <span class="o">=</span> <span class="n">temp_files_path</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>

    <span class="c1"># get list of taxid (labels in fastDNA)</span>
    <span class="c1"># labels = []</span>
    <span class="c1"># for org in filenames:</span>
    <span class="c1">#     try:</span>
    <span class="c1">#         labels.append(&quot;_&quot;.join(re.split(&#39; |\; |\. |\, &#39;, host_data[org][&quot;lineage_names&quot;][-1])))</span>
    <span class="c1">#     except KeyError:</span>
    <span class="c1">#         labels.append(&quot;_&quot;.join(re.split(&#39; |\; |\. |\, &#39;, virus_data[org][&quot;host&quot;][&quot;lineage_names&quot;][-1])))</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; |\; |\. |\, &#39;</span><span class="p">,</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>
    <span class="c1"># labels = [host_data[host][&quot;taxid&quot;] for host in filenames]</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">filepaths</span><span class="p">,</span> <span class="n">labels</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.fasta_parallel">
<code class="highlight language-python"><span class="n">fasta_parallel</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Reads FASTA files and returns parsed data. This function is run in parallel.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>file</code></td>
        <td><code>str</code></td>
        <td><p>Path to FASTA file.</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>SeqRecord</code></td>
      <td><p>Parsed data from FASTA file as a SeqRecord object.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">fasta_parallel</span><span class="p">(</span><span class="n">file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Reads FASTA files and returns parsed data. This function is run in parallel.</span>

<span class="sd">    Args:</span>
<span class="sd">        file: Path to FASTA file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        SeqRecord: Parsed data from FASTA file as a SeqRecord object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">record</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.hybrid_filtering">
<code class="highlight language-python"><span class="n">hybrid_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Edwards' dataset of host is filtered to certain taxonomy level and specific host genomes that represent
taxonomy level are randomly chosen. Additionally, virus genomes associated with each chosen host are also passed
into model training.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>filter</code></td>
        <td><code>str</code></td>
        <td><p>A flag to filter through this function (TODO: this is not final, this will act as an actual filter value in the future)</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>reps</code></td>
        <td><code>int</code></td>
        <td><p>A number of representatives chosen randomly at selected level (defined by <code>filter</code>)</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>Tuple[List[str], List[str]]</code></td>
      <td><p>A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">hybrid_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reps</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;Edwards&#39; dataset of host is filtered to certain taxonomy level and specific host genomes that represent</span>
<span class="sd">    taxonomy level are randomly chosen. Additionally, virus genomes associated with each chosen host are also passed</span>
<span class="sd">    into model training.</span>

<span class="sd">    Args:</span>
<span class="sd">        filter: A flag to filter through this function (TODO: this is not final, this will act as an actual filter value in the future)</span>
<span class="sd">        reps: A number of representatives chosen randomly at selected level (defined by `filter`)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[List[str], List[str]]: A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># filtering all hosts by a chosen level</span>
    <span class="n">filter_host</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">host_data</span><span class="p">:</span>
        <span class="n">level</span> <span class="o">=</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="n">all_tax_levels</span><span class="p">[</span><span class="nb">filter</span><span class="p">][</span><span class="s1">&#39;family&#39;</span><span class="p">]]</span>  <span class="c1"># tax level code</span>
        <span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">host</span><span class="p">)</span>

    <span class="c1"># random sampling of a single host from a level</span>
    <span class="n">random_level_host</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">level</span><span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">])</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">])</span> <span class="o">&lt;</span> <span class="n">reps</span> <span class="k">else</span> <span class="n">reps</span><span class="p">)</span>
        <span class="k">for</span>
        <span class="n">level</span> <span class="ow">in</span> <span class="n">filter_host</span><span class="p">}</span>
    <span class="c1"># print(random_level_host)</span>

    <span class="c1"># list with full paths</span>
    <span class="n">random_level_host_paths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">level</span> <span class="ow">in</span> <span class="n">random_level_host</span><span class="p">:</span>
        <span class="n">random_level_host_paths</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">level</span><span class="p">:</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;HOST&#39;</span><span class="p">][</span><span class="s1">&#39;host_genomes&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">host</span><span class="si">}</span><span class="s2">.fna&quot;</span> <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">random_level_host</span><span class="p">[</span><span class="n">level</span><span class="p">]]})</span>

    <span class="c1"># add known viruses</span>
    <span class="k">for</span> <span class="n">level</span> <span class="ow">in</span> <span class="n">random_level_host</span><span class="p">:</span>
        <span class="n">to_add</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">random_level_host</span><span class="p">[</span><span class="n">level</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># virus_sample = random.sample(hostvir_data[host][&#39;virus_id&#39;], len(random_level_host[level]) if len(random_level_host[level]) &lt; reps else reps)</span>
                <span class="c1">#virus_sample = random.sample(hostvir_data[host][&#39;virus_id&#39;], 1)</span>
                <span class="n">virus_sample</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">hostvir_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s1">&#39;virus_id&#39;</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">hostvir_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s1">&#39;virus_id&#39;</span><span class="p">]))</span>
                <span class="n">to_add</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">virus_sample</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="n">random_level_host_paths</span><span class="p">[</span><span class="n">level</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;VIRUS&#39;</span><span class="p">][</span><span class="s1">&#39;virus_genomes&#39;</span><span class="p">]</span><span class="si">}{</span><span class="n">virus</span><span class="si">}</span><span class="s2">.fna&quot;</span> <span class="k">for</span> <span class="n">virus</span> <span class="ow">in</span> <span class="n">to_add</span><span class="p">])</span>
        <span class="n">random_level_host</span><span class="p">[</span><span class="n">level</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">to_add</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">random_level_host</span><span class="p">)</span>

    <span class="c1"># list of filenames to be used</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">random_level_host</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="c1"># print(filenames)</span>

    <span class="c1"># list of paths to be returned</span>
    <span class="n">filepaths</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">random_level_host_paths</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="c1"># flatten if needed</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filenames</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">filenames</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">temp_files_path</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">filepaths</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="n">temp_files</span>
        <span class="n">filepaths</span> <span class="o">=</span> <span class="n">temp_files_path</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>

    <span class="c1"># get list of taxid (labels in fastDNA)</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">org</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; |\; |\. |\, &#39;</span><span class="p">,</span> <span class="n">host_data</span><span class="p">[</span><span class="n">org</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">])))</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; |\; |\. |\, &#39;</span><span class="p">,</span> <span class="n">virus_data</span><span class="p">[</span><span class="n">org</span><span class="p">][</span><span class="s2">&quot;host&quot;</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">])))</span>
    <span class="c1">#labels = [&quot;_&quot;.join(re.split(&#39; |\; |\. |\, &#39;, host_data[host][&quot;lineage_names&quot;][-1])) for host in filenames]</span>
    <span class="c1"># labels = [host_data[host][&quot;taxid&quot;] for host in filenames]</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">filepaths</span><span class="p">,</span> <span class="n">labels</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.main_procedure">
<code class="highlight language-python"><span class="n">main_procedure</span><span class="p">(</span><span class="n">input_dir</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="nb">filter</span><span class="p">,</span> <span class="n">dim</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">minn</span><span class="p">,</span> <span class="n">maxn</span><span class="p">,</span> <span class="n">epoch</span><span class="p">,</span> <span class="n">thread</span><span class="p">,</span> <span class="n">reps</span><span class="p">,</span> <span class="n">rm</span><span class="p">,</span> <span class="n">save_vec</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Main function of the module that runs model making.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>input_dir</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory with all host genomes.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>out_dir</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory where all output files are saved (chosen genomes - <code>.fasta</code> file, taxid list - <code>.txt</code> file, model file - <code>.bin</code> file)</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>filter</code></td>
        <td><code>str</code></td>
        <td><p>A taxonomy level by which filtering is conducted / A flag to filter through this function</p>
<div class="admonition todo">
<p class="admonition-title">Todo</p>
</div>
<p><code>filter</code> argument is ambiguous at this moment. This is not final, this will act as an actual filter value in the future</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>dim</code></td>
        <td><code>int</code></td>
        <td><p>Dimensionality of vectors to be used during the training process.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>length</code></td>
        <td><code>int</code></td>
        <td><p>Length of fragments of the genomes to be used during the training process.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>minn</code></td>
        <td><code>int</code></td>
        <td><p>Minimum size of a k-mer to be used during the training process.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>maxn</code></td>
        <td><code>int</code></td>
        <td><p>Maximum size of a k-mer to be used during the training process (it is advised to be kept the same as <code>minn</code> and <strong>less than 15, otherwise fastDNA fails</strong>).</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>epoch</code></td>
        <td><code>int</code></td>
        <td><p>Number of training epochs (each added epoch increases runtime significantly, but in most cases increases model quality).</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>thread</code></td>
        <td><code>int</code></td>
        <td><p>Number of CPU threads to be used during training process (more threads - faster training, but higher CPU usage)</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>reps</code></td>
        <td><code>int</code></td>
        <td><p>A number of representatives chosen randomly at selected level (defined by <code>filter</code>)</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>rm</code></td>
        <td><code>bool</code></td>
        <td><p>Remove potentially redundant files after model creation. Default is <code>false</code> and it is advised to be set default.</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>save_vec</code></td>
        <td><code>bool</code></td>
        <td><p>Enables saving of a readable model file <code>.vec</code>. Enabling this may significantly increase execution time. Default is 'false'.</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>      <div class="admonition todo">
<p class="admonition-title">Todo</p>
<p><code>filter</code> argument is ambiguous at this moment. This is not final, this will act as an actual filter value in the future</p>
</div>

        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">main_procedure</span><span class="p">(</span><span class="n">input_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">filter</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dim</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">minn</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">maxn</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">epoch</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                   <span class="n">thread</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">reps</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">rm</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">save_vec</span><span class="p">:</span> <span class="nb">bool</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Main function of the module that runs model making.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_dir: Path to directory with all host genomes.</span>
<span class="sd">        out_dir: Path to directory where all output files are saved (chosen genomes - `.fasta` file, taxid list - `.txt` file, model file - `.bin` file)</span>
<span class="sd">        filter: A taxonomy level by which filtering is conducted / A flag to filter through this function</span>
<span class="sd">            !!! todo</span>
<span class="sd">            `filter` argument is ambiguous at this moment. This is not final, this will act as an actual filter value in the future</span>
<span class="sd">        dim: Dimensionality of vectors to be used during the training process.</span>
<span class="sd">        length: Length of fragments of the genomes to be used during the training process.</span>
<span class="sd">        minn: Minimum size of a k-mer to be used during the training process.</span>
<span class="sd">        maxn: Maximum size of a k-mer to be used during the training process (it is advised to be kept the same as `minn` and **less than 15, otherwise fastDNA fails**).</span>
<span class="sd">        epoch: Number of training epochs (each added epoch increases runtime significantly, but in most cases increases model quality).</span>
<span class="sd">        thread: Number of CPU threads to be used during training process (more threads - faster training, but higher CPU usage)</span>
<span class="sd">        reps: A number of representatives chosen randomly at selected level (defined by `filter`)</span>
<span class="sd">        rm: Remove potentially redundant files after model creation. Default is `false` and it is advised to be set default.</span>
<span class="sd">        save_vec: Enables saving of a readable model file `.vec`. Enabling this may significantly increase execution time. Default is &#39;false&#39;.</span>

<span class="sd">    !!! todo</span>

<span class="sd">        `filter` argument is ambiguous at this moment. This is not final, this will act as an actual filter value in the future</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># colorama</span>
    <span class="n">init</span><span class="p">()</span>

    <span class="c1"># timeit</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">timer</span><span class="p">()</span>

    <span class="n">out_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">out_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;mkdir </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">out_path</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="nb">filter</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="s2">&quot;hybrid&quot;</span><span class="p">,</span> <span class="s2">&quot;debug&quot;</span><span class="p">]:</span>
        <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">tax_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">filter</span> <span class="o">==</span> <span class="s2">&quot;none&quot;</span><span class="p">:</span>
        <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">no_filtering</span><span class="p">(</span><span class="n">input_dir</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">filter</span> <span class="o">==</span> <span class="s2">&quot;hybrid&quot;</span><span class="p">:</span>
        <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">hybrid_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">filter</span> <span class="o">==</span> <span class="s2">&quot;debug&quot;</span><span class="p">:</span>
        <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">debug_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span>

    <span class="c1"># parse selected files and merge them into single fasta file; create labels file</span>
    <span class="c1"># records = [list(SeqIO.parse(f&quot;D:/praktyki2020/edwards2016/host/fasta/{file}.fna&quot;, &quot;fasta&quot;))[0] for file in filenames]</span>
    <span class="c1"># for file, label in zip(filenames, labels):</span>
    <span class="c1">#     print(f&quot;{file} - {label}&quot;)</span>
    <span class="k">if</span> <span class="nb">filter</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;hybrid&quot;</span><span class="p">,</span> <span class="s2">&quot;debug&quot;</span><span class="p">]:</span>
        <span class="n">par</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">11</span><span class="p">,</span> <span class="n">pre_dispatch</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">fasta_parallel</span><span class="p">)(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_dir</span><span class="si">}{</span><span class="n">file</span><span class="si">}</span><span class="s2">.fna&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">filter</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;hybrid&#39;</span><span class="p">,</span> <span class="s2">&quot;debug&quot;</span><span class="p">]:</span>
        <span class="n">par</span> <span class="o">=</span> <span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">11</span><span class="p">,</span> <span class="n">pre_dispatch</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">backend</span><span class="o">=</span><span class="s2">&quot;loky&quot;</span><span class="p">)(</span>
            <span class="n">delayed</span><span class="p">(</span><span class="n">fasta_parallel</span><span class="p">)(</span><span class="n">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">)</span>
    <span class="c1"># print(par)</span>
    <span class="c1"># print(len(records))</span>
    <span class="c1"># print(records)</span>

    <span class="n">filtered_fasta_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;random_train-</span><span class="si">{</span><span class="nb">filter</span><span class="si">}</span><span class="s2">-dim_</span><span class="si">{</span><span class="n">dim</span><span class="si">}</span><span class="s2">-len_</span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2">.fasta&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">filtered_fasta_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w_fh</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">par</span><span class="p">,</span> <span class="n">w_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

    <span class="n">labels_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;random_labels-</span><span class="si">{</span><span class="nb">filter</span><span class="si">}</span><span class="s2">-dim_</span><span class="si">{</span><span class="n">dim</span><span class="si">}</span><span class="s2">-len_</span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2">.txt&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">labels_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">label</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># run fastDNA training</span>
    <span class="n">model_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;random_model-</span><span class="si">{</span><span class="nb">filter</span><span class="si">}</span><span class="s2">-dim_</span><span class="si">{</span><span class="n">dim</span><span class="si">}</span><span class="s2">-len_</span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2">-epoch</span><span class="si">{</span><span class="n">epoch</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">if</span> <span class="n">save_vec</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;GENERAL&#39;</span><span class="p">][</span><span class="s1">&#39;fastdna_dir&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">fastdna supervised -input </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">filtered_fasta_file</span><span class="si">}</span><span class="s2"> -labels </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">labels_file</span><span class="si">}</span><span class="s2"> -output </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">model_file</span><span class="si">}</span><span class="s2"> -dim </span><span class="si">{</span><span class="n">dim</span><span class="si">}</span><span class="s2"> -length </span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2"> -minn </span><span class="si">{</span><span class="n">minn</span><span class="si">}</span><span class="s2"> -maxn </span><span class="si">{</span><span class="n">maxn</span><span class="si">}</span><span class="s2"> -epoch </span><span class="si">{</span><span class="n">epoch</span><span class="si">}</span><span class="s2"> -thread </span><span class="si">{</span><span class="n">thread</span><span class="si">}</span><span class="s2"> -saveVec&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;GENERAL&#39;</span><span class="p">][</span><span class="s1">&#39;fastdna_dir&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">fastdna supervised -input </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">filtered_fasta_file</span><span class="si">}</span><span class="s2"> -labels </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">labels_file</span><span class="si">}</span><span class="s2"> -output </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">model_file</span><span class="si">}</span><span class="s2"> -dim </span><span class="si">{</span><span class="n">dim</span><span class="si">}</span><span class="s2"> -length </span><span class="si">{</span><span class="n">length</span><span class="si">}</span><span class="s2"> -minn </span><span class="si">{</span><span class="n">minn</span><span class="si">}</span><span class="s2"> -maxn </span><span class="si">{</span><span class="n">maxn</span><span class="si">}</span><span class="s2"> -epoch </span><span class="si">{</span><span class="n">epoch</span><span class="si">}</span><span class="s2"> -thread </span><span class="si">{</span><span class="n">thread</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># par = Parallel(n_jobs=-1)(delayed(SeqIO.write)(records, &quot;D:/praktyki2020/edwards2016/host/random_phylum-training_fastDNA.fasta&quot;, &quot;fasta&quot;) for record in records)</span>
    <span class="c1"># SeqIO.write(records, &quot;X:/edwards2016/host/random_phylum-training_fastDNA.fasta&quot;, &quot;fasta&quot;)</span>
    <span class="c1"># print(f&quot;Written {len(records)} records.&quot;)</span>

    <span class="k">if</span> <span class="n">rm</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;removing&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;rm </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">filtered_fasta_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;rm </span><span class="si">{</span><span class="n">out_dir</span><span class="si">}{</span><span class="n">labels_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">end</span> <span class="o">=</span> <span class="n">timer</span><span class="p">()</span>
    <span class="n">runtime</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="n">start</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Fore</span><span class="o">.</span><span class="n">GREEN</span><span class="si">}</span><span class="s2">[make-model] Done in </span><span class="si">{</span><span class="n">runtime</span><span class="si">:</span><span class="s2">.6f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.no_filtering">
<code class="highlight language-python"><span class="n">no_filtering</span><span class="p">(</span><span class="n">input_dir</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Edwards' dataset of host is not filtered and all genomes from selected directory are passed to model training.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>input_dir</code></td>
        <td><code>str</code></td>
        <td><p>Path to directory with all host genomes.</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>Tuple[List[str], List[str]]</code></td>
      <td><p>A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">no_filtering</span><span class="p">(</span><span class="n">input_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;Edwards&#39; dataset of host is not filtered and all genomes from selected directory are passed to model training.</span>

<span class="sd">        Args:</span>
<span class="sd">            input_dir: Path to directory with all host genomes.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[List[str], List[str]]: A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">input_dir</span><span class="si">}</span><span class="s2">*.fna&quot;</span><span class="p">)]</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>
    <span class="c1"># labels = [host_data[host][&quot;taxid&quot;] for host in filenames]</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; |\; |\. |\, &#39;</span><span class="p">,</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span>
</code></pre></div>
        </details>
    </div>

  </div>



  <div class="doc doc-object doc-function">



<h2 class="doc doc-heading" id="make-model.tax_filtering">
<code class="highlight language-python"><span class="n">tax_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="n">reps</span><span class="p">)</span></code>


</h2>

    <div class="doc doc-contents ">

      <p>Edwards' dataset of host is filtered to certain taxonomy level and specific host genomes that represent
taxonomy level are randomly chosen.</p>

<p><strong>Parameters:</strong></p>
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
    </tr>
  </thead>
  <tbody>
      <tr>
        <td><code>filter</code></td>
        <td><code>str</code></td>
        <td><p>A taxonomy level by which filtering is conducted</p></td>
        <td><em>required</em></td>
      </tr>
      <tr>
        <td><code>reps</code></td>
        <td><code>int</code></td>
        <td><p>A number of representatives chosen randomly at selected level (defined by <code>filter</code>)</p></td>
        <td><em>required</em></td>
      </tr>
  </tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
  <thead>
    <tr>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>Tuple[List[str], List[str]]</code></td>
      <td><p>A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</p></td>
    </tr>
  </tbody>
</table>
        <details class="quote">
          <summary>Source code in <code>fastdna-faiss\make-model.py</code></summary>
          <div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">tax_filtering</span><span class="p">(</span><span class="nb">filter</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reps</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;Edwards&#39; dataset of host is filtered to certain taxonomy level and specific host genomes that represent</span>
<span class="sd">    taxonomy level are randomly chosen.</span>

<span class="sd">    Args:</span>
<span class="sd">        filter: A taxonomy level by which filtering is conducted</span>
<span class="sd">        reps: A number of representatives chosen randomly at selected level (defined by `filter`)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[List[str], List[str]]: A list of ncbi_id of selected hosts genomes and a list of taxid of selected hosts.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># filtering all hosts by a chosen level</span>
    <span class="n">filter_host</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">host_data</span><span class="p">:</span>
        <span class="n">level</span> <span class="o">=</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="n">all_tax_levels</span><span class="p">[</span><span class="nb">filter</span><span class="p">]]</span>  <span class="c1"># tax level code</span>
        <span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">host</span><span class="p">)</span>

    <span class="c1"># random sampling of a single host from a level</span>
    <span class="n">random_level_host</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">level</span><span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">])</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_host</span><span class="p">[</span><span class="n">level</span><span class="p">])</span> <span class="o">&lt;</span> <span class="n">reps</span> <span class="k">else</span> <span class="n">reps</span><span class="p">)</span>
        <span class="k">for</span>
        <span class="n">level</span> <span class="ow">in</span> <span class="n">filter_host</span><span class="p">}</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">random_level_host</span><span class="p">)</span>

    <span class="c1"># list of filenames to be used</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">random_level_host</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="c1"># print(filenames)</span>

    <span class="c1"># flatten if needed</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">filenames</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">filenames</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="n">temp_files</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>

    <span class="c1"># get list of taxid (labels in fastDNA)</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; |\; |\. |\, &#39;</span><span class="p">,</span> <span class="n">host_data</span><span class="p">[</span><span class="n">host</span><span class="p">][</span><span class="s2">&quot;lineage_names&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">for</span> <span class="n">host</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">]</span>
    <span class="c1"># labels = [host_data[host][&quot;taxid&quot;] for host in filenames]</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">filenames</span><span class="p">,</span> <span class="n">labels</span>
</code></pre></div>
        </details>
    </div>

  </div>






  </div>

    </div>

  </div>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../.." class="md-footer__link md-footer__link--prev" aria-label="Previous: Main page" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Main page
            </div>
          </div>
        </a>
      
      
        
        <a href="../random_sampling/" class="md-footer__link md-footer__link--next" aria-label="Next: random_sampling" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              random_sampling
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        
          Made with
          <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
            Material for MkDocs
          </a>
        
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.e99e714c.min.js", "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.bc35569b.min.js"></script>
      
        <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
      
    
  </body>
</html>